processor/samplingprocessor/tailsamplingprocessor/processor.go

```diff
@@ -152,66 +152,87 @@ func (tsp *tailSamplingSpanProcessor) samplingPolicyOnTick() {
 		}
 		trace := d.(*sampling.TraceData)
 		trace.DecisionTime = time.Now()
-		for i, policy := range tsp.policies {
-			policyEvaluateStartTime := time.Now()
-			decision, err := policy.Evaluator.Evaluate(id, trace)
-			stats.Record(
-				policy.ctx,
-				statDecisionLatencyMicroSec.M(int64(time.Since(policyEvaluateStartTime)/time.Microsecond)))
-			if err != nil {
-				trace.Decisions[i] = sampling.NotSampled
-				evaluateErrorCount++
-				tsp.logger.Error("Sampling policy error", zap.Error(err))
-				continue
+
+		decision, policy, sampleCount, notSampleCount, errorCount := tsp.makeDecision(id, trace)
+		decisionSampled += sampleCount
+		decisionNotSampled += notSampleCount
+		evaluateErrorCount += errorCount
+
+		// Sampled or not, remove the batches
+		trace.Lock()
+		traceBatches := trace.ReceivedBatches
+		trace.ReceivedBatches = nil
+		trace.Unlock()
+
+		if decision == sampling.Sampled {
+			for j := 0; j < len(traceBatches); j++ {
+				_ = tsp.nextConsumer.ConsumeTraces(policy.ctx, internaldata.OCToTraceData(traceBatches[j]))
 			}
+		}
+	}
+
+	stats.Record(tsp.ctx,
+		statOverallDecisionLatencyµs.M(int64(time.Since(startTime)/time.Microsecond)),
+		statDroppedTooEarlyCount.M(idNotFoundOnMapCount),
+		statPolicyEvaluationErrorCount.M(evaluateErrorCount),
+		statTracesOnMemoryGauge.M(int64(atomic.LoadUint64(&tsp.numTracesOnMap))))
+
+	tsp.logger.Debug("Sampling policy evaluation completed",
+		zap.Int("batch.len", batchLen),
+		zap.Int64("sampled", decisionSampled),
+		zap.Int64("notSampled", decisionNotSampled),
+		zap.Int64("droppedPriorToEvaluation", idNotFoundOnMapCount),
+		zap.Int64("policyEvaluationErrors", evaluateErrorCount),
+	)
+}
+
+func (tsp *tailSamplingSpanProcessor) makeDecision(id idbatcher.ID, trace *sampling.TraceData) (sampling.Decision, *Policy, int64, int64, int64) {
+	var decisionSampled, decisionNotSampled, evaluateErrorCount int64
+	finalDecision := sampling.NotSampled
+	var matchingPolicy *Policy = nil
+
+	for i, policy := range tsp.policies {
+		policyEvaluateStartTime := time.Now()
+		decision, err := policy.Evaluator.Evaluate(id, trace)
+		stats.Record(
+			policy.ctx,
+			statDecisionLatencyMicroSec.M(int64(time.Since(policyEvaluateStartTime)/time.Microsecond)))
 
+		if err != nil {
+			trace.Decisions[i] = sampling.NotSampled
+			evaluateErrorCount++
+			tsp.logger.Error("Sampling policy error", zap.Error(err))
+		} else {
 			trace.Decisions[i] = decision
 
 			switch decision {
 			case sampling.Sampled:
-				stats.RecordWithTags(
+				// any single policy that decides to sample will cause the decision to be sampled
+				// the nextConsumer will get the context from the first matching policy
+				finalDecision = sampling.Sampled
+				if matchingPolicy == nil {
```

Comment:
The only reason this is necessary is to get the matching context from the policy which contains tags for the policy name